plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryQueryPageSource.java

```diff
@@ -0,0 +1,246 @@
+import static io.trino.spi.type.Timestamps.NANOSECONDS_PER_DAY;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_DAY;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_NANOSECOND;
+import static io.trino.spi.type.Timestamps.round;
+import static java.lang.String.format;
+import static java.time.temporal.ChronoField.NANO_OF_SECOND;
+import static java.util.Objects.requireNonNull;
+
+public class BigQueryQueryPageSource
+        implements ConnectorPageSource
+{
+    private static final DateTimeFormatter TIME_FORMATTER = new DateTimeFormatterBuilder()
+            .appendPattern("HH:mm:ss")
+            .optionalStart()
+            .appendFraction(NANO_OF_SECOND, 0, 6, true)
+            .optionalEnd()
+            .toFormatter();
+
+    private final List<Type> columnTypes;
+    private final PageBuilder pageBuilder;
+    private final TableResult tableResult;
+
+    private boolean finished;
+
+    public BigQueryQueryPageSource(
+            BigQueryClient client,
+            BigQueryTableHandle table,
+            List<String> columnNames,
+            List<Type> columnTypes,
+            Optional<String> filter)
+    {
+        requireNonNull(client, "client is null");
+        requireNonNull(table, "table is null");
+        requireNonNull(columnNames, "columnNames is null");
+        requireNonNull(columnTypes, "columnTypes is null");
+        requireNonNull(filter, "filter is null");
+        checkArgument(columnNames.size() == columnTypes.size(), "columnNames and columnTypes sizes don't match");
+        this.columnTypes = ImmutableList.copyOf(columnTypes);
+        this.pageBuilder = new PageBuilder(columnTypes);
+        TableId tableId = TableId.of(client.getProjectId(), table.getRemoteTableName().getDatasetName(), table.getRemoteTableName().getTableName());
+        this.tableResult = client.query(tableId, ImmutableList.copyOf(columnNames), filter);
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished;
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        verify(pageBuilder.isEmpty());
+        for (FieldValueList record : tableResult.iterateAll()) {
+            pageBuilder.declarePosition();
+            for (int column = 0; column < columnTypes.size(); column++) {
+                BlockBuilder output = pageBuilder.getBlockBuilder(column);
+                appendTo(columnTypes.get(column), record.get(column), output);
+            }
+        }
+        finished = true;
+
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
+        return page;
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        return 0;
+    }
+
+    private void appendTo(Type type, FieldValue value, BlockBuilder output)
```

Comment:
A TODO comment + issue about what types are unsupported (BIGNUMERIC, TIMESTAMP WITH TIME ZONE?) where we throw would be useful.

Should it also be documented?